### PR TITLE
zgui: expose color utilities

### PR DESCRIPTION
**Pull request by Jack-Ji**
1. Rearranged tables api
2. Expose color utility 
3. Update zgui sample to use color utility to get color value

### CODE REVIEW

File: libs/zgui/src/gui.zig [hunk 1]
```diff
@@ -2680,6 +2913,42 @@ extern fn zguiIsAnyItemActive() bool;
...
+
+pub fn colorConvertFloat4ToU32(in: [4]f32) u32 {
+    return zguiColorConvertFloat4ToU32(&in);
+}
+
+pub fn colorConvertFloat3ToU32(in: [3]f32) u32 {
+    return colorConvertFloat4ToU32(.{ in[0], in[1], in[2], 1 });
+}
+
+pub fn colorConvertRGBtoHSV(r: f32, g: f32, b: f32) [3]f32 {
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:2936]: Let's use `colorConvertRgbToHsv` name.

File: libs/zgui/src/gui.zig [hunk 2]
```diff
@@ -2680,6 +2913,42 @@ extern fn zguiIsAnyItemActive() bool;
...
+    var hsv: [3]f32 = undefined;
+    return zguiColorConvertRGBtoHSV(r, g, b, &hsv[0], &hsv[1], &hsv[2]);
+}
+
+pub fn colorConvertHSVtoRGB(h: f32, s: f32, v: f32) [3]f32 {
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:2941]: Let's use `colorConvertHsvToRgb` name.

File: libs/zgui/src/zgui.cpp [hunk 3]
```diff
@@ -1463,6 +1463,27 @@ ZGUI_API void zguiTableSetBgColor(ImGuiTableBgTarget target, unsigned int color,
...
+//--------------------------------------------------------------------------------------------------
+//
+// Color Utilities
+//
+//--------------------------------------------------------------------------------------------------
+ZGUI_API ImVec4 zguiColorConvertU32ToFloat4(ImU32 in) {
+    return ImGui::ColorConvertU32ToFloat4(in);
+}
+
+ZGUI_API ImU32 zguiColorConvertFloat4ToU32(const ImVec4 in) {
```
Comment (michal-z) [inline libs/zgui/src/zgui.cpp:1475]:
```suggestion
ZGUI_API ImU32 zguiColorConvertFloat4ToU32(const float color[4]) {
```

**Review by michal-z — CHANGES_REQUESTED** [review]
Thanks! I've marked few tiny issues.

**Comment by Jack-Ji** [comment]
It's done. Please check again.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "michal-z/zig-gamedev"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年12月7日(星期三) 晚上6:35
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [michal-z/zig-gamedev] zgui: expose color utilities (PR #168)





 
@michal-z requested changes on this pull request.
 
Thanks! I've marked few tiny issues.

 
In libs/zgui/src/gui.zig:
 &gt; +} + +pub fn colorConvertU32ToFloat3(in: u32) [3]f32 { +    const rgba = zguiColorConvertU32ToFloat4(in); +    return .{ rgba[0], rgba[1], rgba[2] }; +} + +pub fn colorConvertFloat4ToU32(in: [4]f32) u32 { +    return zguiColorConvertFloat4ToU32(&amp;in); +} + +pub fn colorConvertFloat3ToU32(in: [3]f32) u32 { +    return colorConvertFloat4ToU32(.{ in[0], in[1], in[2], 1 }); +} + +pub fn colorConvertRGBtoHSV(r: f32, g: f32, b: f32) [3]f32 {  
Let's use colorConvertRgbToHsv name.
 
 
In libs/zgui/src/gui.zig:
 &gt; +} + +pub fn colorConvertFloat4ToU32(in: [4]f32) u32 { +    return zguiColorConvertFloat4ToU32(&amp;in); +} + +pub fn colorConvertFloat3ToU32(in: [3]f32) u32 { +    return colorConvertFloat4ToU32(.{ in[0], in[1], in[2], 1 }); +} + +pub fn colorConvertRGBtoHSV(r: f32, g: f32, b: f32) [3]f32 { +    var hsv: [3]f32 = undefined; +    return zguiColorConvertRGBtoHSV(r, g, b, &amp;hsv[0], &amp;hsv[1], &amp;hsv[2]); +} + +pub fn colorConvertHSVtoRGB(h: f32, s: f32, v: f32) [3]f32 {  
Let's use colorConvertHsvToRgb name.
 
 
In libs/zgui/src/zgui.cpp:
 &gt; @@ -1463,6 +1463,27 @@ ZGUI_API void zguiTableSetBgColor(ImGuiTableBgTarget target, unsigned int color,      ImGui::TableSetBgColor(target, color, column_n);  }   +//-------------------------------------------------------------------------------------------------- +// +// Color Utilities +// +//-------------------------------------------------------------------------------------------------- +ZGUI_API ImVec4 zguiColorConvertU32ToFloat4(ImU32 in) { +    return ImGui::ColorConvertU32ToFloat4(in); +} + +ZGUI_API ImU32 zguiColorConvertFloat4ToU32(const ImVec4 in) {  ⬇️ Suggested change -ZGUI_API ImU32 zguiColorConvertFloat4ToU32(const ImVec4 in) { +ZGUI_API ImU32 zguiColorConvertFloat4ToU32(const float color[4]) {

**Pull request merged**